src/Mod/Path/Path/Op/Custom.py

```diff
@@ -47,21 +49,77 @@ def opFeatures(self, obj):
         return PathOp.FeatureTool | PathOp.FeatureCoolant
 
     def initOperation(self, obj):
+        obj.addProperty(
+            "App::PropertyEnumeration",
+            "Source",
+            "Path",
+            "Source of gcode (text, file, ...)"
+        )
+
+        obj.addProperty(
+            "App::PropertyFile",
+            "GcodeFile",
+            "Path",
+            "File containing gcode to be inserted",
+        )
+
         obj.addProperty(
             "App::PropertyStringList",
             "Gcode",
             "Path",
             QT_TRANSLATE_NOOP("App::Property", "The gcode to be inserted"),
         )
 
+        obj.Source = [ "Text", "File" ]
         obj.Proxy = self
+        self.setEditorModes(obj)
+
+    def onChanged(self, obj, prop):
+        if prop == "Source":
+            self.setEditorModes(obj)
+
+    def onDocumentRestore(self, obj):
+        self.setEditorModes(self, obj)
+
+    def setEditorModes(self, obj, features=None):
+        if not hasattr(obj, "Source"):
+            return
+
+        if obj.Source == "Text":
+          obj.setEditorMode("GcodeFile", 2)
+          obj.setEditorMode("Gcode", 0)
+        elif obj.Source == "File":
+          obj.setEditorMode("GcodeFile", 0)
+          obj.setEditorMode("Gcode", 2)
+
+    def findGcodeFile(self, filename):
+        if os.path.exists(filename):
+            # probably absolute, just return
+            return filename
+
+        doc_path = os.path.dirname(FreeCAD.ActiveDocument.FileName)
+        prospective_path = os.path.join(doc_path, filename)
+
+        if os.path.exists(prospective_path):
+            return prospective_path
 
     def opExecute(self, obj):
         self.commandlist.append(Path.Command("(Begin Custom)"))
-        if obj.Gcode:
+
+        if obj.Source == "Text" and obj.Gcode:
             for l in obj.Gcode:
                 newcommand = Path.Command(str(l))
                 self.commandlist.append(newcommand)
+        elif obj.Source == "File" and len(obj.GcodeFile) > 0:
+            gcode_file = self.findGcodeFile(obj.GcodeFile)
+
+            # could not determine the path
+            if not gcode_file: return
+
+            with open(gcode_file) as fd:
+                for l in fd.readlines():
+                  newcommand = Path.Command(str(l))
```

Comment:
This worries me a bit, FreeCAD has a fairly strict definition on what GCode Ops are supported and in what format (line numbers/comments/etc...).  When a custom op is normally created from a file, it uses the behavior in Post/scripts/gcode_pre.py to only include the "whitelisted" functionality that the post_processors can handle.

On the other hand, it is allowed to put any custom gcode in here regardless.  It just may have some odd side effects.

I guess my main concern is that it differs from the "standard" import workflow.

Reply:
So far I've been approaching this with the goal of minimizing changes in default behavior so I just copied how it was done previously. I guess the bigger picture question is, **should all internal gcode be parsed to this standard format?**

This could also be put behind a Property switch to enable strict behavior without changing the default.

Reply:
It's a tricky thing to answer.  Right now gcode_pre is *very* limited, but that's by design.  You are correct that it's a bigger question of how we should handle this scenario in general.

CustomOp is currently loaded with foot-guns, maybe we keep that for now and don't worry about it here.